File: java/src/main/java/com/workshop/architecture/fitness/application/MembershipService.java
package com.workshop.architecture.fitness.application;

import com.workshop.architecture.fitness.application.port.inbound.ActivateMembership;
import com.workshop.architecture.fitness.application.port.outbound.*;
import com.workshop.architecture.fitness.domain.InvoiceDetails;
import com.workshop.architecture.fitness.domain.Membership;
import com.workshop.architecture.fitness.domain.MembershipBillingReference;
import com.workshop.architecture.fitness.infrastructure.*;
import com.workshop.architecture.fitness.infrastructure.external_invoice_provider.ExternalInvoiceProviderClient;
import jakarta.transaction.Transactional;
import org.jspecify.annotations.NonNull;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.LocalDate;
import java.time.Period;
import java.util.UUID;

@Service
public class MembershipService implements ActivateMembership {
    private final MembershipRepository membershipRepository;
    private final PlanRepository planRepository;
    private final InMemoryEmailService emailService;
    private final String billingSenderEmailAddress;
    private final ForFindingCustomers forFindingCustomers;
    private final ForFindingPlans forFindingPlans;
    private final ForStoringMemberships forStoringMemberships;
    private final ForCreatingInvoices forCreatingInvoices;
    private final ForStoringBillingReferences forStoringBillingReferences;
    private final ForSendingEmails forSendingEmails;

    private final MembershipBillingReferenceRepository billingReferenceRepository;
    private final ExternalInvoiceProviderClient externalInvoiceProviderClient;

    public MembershipService(
            MembershipRepository membershipRepository,
            CustomerRepository customerRepository,
            PlanRepository planRepository,
            InMemoryEmailService emailService,
            @Value("${workshop.billing.sender-email-address}") String billingSenderEmailAddress,
            ForFindingCustomers forFindingCustomers,
            ForFindingPlans forFindingPlans,
            ForStoringMemberships forStoringMemberships,
            ForCreatingInvoices forCreatingInvoices,
            ForStoringBillingReferences forStoringBillingReferences,
            ForSendingEmails forSendingEmails,
            MembershipBillingReferenceRepository billingReferenceRepository,
            ExternalInvoiceProviderClient externalInvoiceProviderClient
    ) {
        this.membershipRepository = membershipRepository;
        this.planRepository = planRepository;
        this.emailService = emailService;
        this.billingSenderEmailAddress = billingSenderEmailAddress;
        this.forFindingCustomers = forFindingCustomers;
        this.forFindingPlans = forFindingPlans;
        this.forStoringMemberships = forStoringMemberships;
        this.forCreatingInvoices = forCreatingInvoices;
        this.forStoringBillingReferences = forStoringBillingReferences;
        this.forSendingEmails = forSendingEmails;
        this.billingReferenceRepository = billingReferenceRepository;
        this.externalInvoiceProviderClient = externalInvoiceProviderClient;
    }

    @Transactional
    @Override
    public ActivateMembershipResult activateMembership(ActivateMembershipInput input) {

        var customerId = UUID.fromString(input.customerId());
        var customer = forFindingCustomers.findCustomerById(customerId);

        var planId = UUID.fromString(input.planId());
        var plan = forFindingPlans.findPlanById(planId);

        var startDate = LocalDate.now();
        var endDate = startDate.plusMonths(plan.durationInMonths());

        if (Period.between(customer.dateOfBirth(), startDate).getYears() < 18
                && !Boolean.TRUE.equals(input.signedByCustodian())) {
            throw new CustomerTooYoungException(
                    "Customers younger than 18 require signedByCustodian=true"
            );
        }

        var membership = new Membership(
                UUID.randomUUID(),
                customerId,
                planId,
                plan.price().intValue(),
                plan.durationInMonths(),
                "ACTIVE",
                null,
                startDate,
                endDate
        );

        var storedMembership = forStoringMemberships.storeMembership(membership);
        var invoiceDetails = new InvoiceDetails(
                UUID.randomUUID(),
                customerId,
                membership.id(),
                plan.id(),
                LocalDate.now().plusDays(30),
                plan.title(),
                plan.price());

        var externalInvoiceId = forCreatingInvoices.createInvoiceWith(invoiceDetails);


        var now = Instant.now();
        var billingReference = new MembershipBillingReference(
                UUID.randomUUID(),
                storedMembership.id(),
                externalInvoiceId,
                invoiceDetails.membershipId().toString(),
                invoiceDetails.dueDate(),
                "OPEN",
                now,
                now
        );


        var storedBillingReference = forStoringBillingReferences.storeMembershipBillingReference(billingReference);

        sendEmail(toEmail(storedBillingReference, customer.emailAddress(), storedMembership.planPrice()));

        return new ActivateMembershipResult(
                storedBillingReference.membershipId().toString(),
                storedMembership.customerId().toString(),
                storedMembership.planId().toString(),
                storedMembership.planPrice(),
                storedMembership.planDurationInMonths(),
                storedMembership.status(),
                storedMembership.startDate(),
                storedMembership.endDate(),
                storedBillingReference.externalInvoiceReference(),
                storedBillingReference.externalInvoiceId(),
                storedBillingReference.dueDate()
        );
    }


    private static @NonNull CustomerActivateMembershipEmail toEmail(MembershipBillingReference billingReference, String emailAddress, int membershipPlanPrice) {
        return new CustomerActivateMembershipEmail(
                billingReference.externalInvoiceReference(),
                billingReference.dueDate(),
                emailAddress,
                membershipPlanPrice,
                """
                        |
                        |To: %s
                        |From: %s
                        |Subject: Your Membership Invoice %s
                        |
                        |Dear customer,
                        |
                        |Thank you for your membership.
                        |
                        |Please find your invoice details below:
                        |Invoice ID: %s
                        |Amount Due: CHF %s
                        |Due Date: %s
                        |
                        |Attachment: invoice-%s.pdf
                        |
                        |Kind regards,
                        |Codeartify Billing
                        |
                        """);
    }

    private void sendEmail(CustomerActivateMembershipEmail emailDetails) {
        var activationEmail = emailDetails.emailTemplate().formatted(
                        emailDetails.emailAddress(),
                        billingSenderEmailAddress,
                        emailDetails.invoiceId(),
                        emailDetails.invoiceId(),
                        emailDetails.planPrice(),
                        emailDetails.invoiceDueDate(),
                        emailDetails.invoiceId()
                )
                .replace("\n|", "\n")
                .trim();

        System.out.println(activationEmail);

        emailService.send(activationEmail);
    }

}

File: python/src/workshop_api/fitness/business/application/membership_service.py
import calendar
import uuid
from datetime import UTC, date, datetime, timedelta
from decimal import Decimal

from workshop_api.fitness.business.application.activate_membership_input import (
    ActivateMembershipInput,
)
from workshop_api.fitness.business.application.activate_membership_result import (
    ActivateMembershipResult,
)
from workshop_api.fitness.business.application.customer_too_young_exception import (
    CustomerTooYoungException,
)
from workshop_api.fitness.business.domain.customer_activate_membership_email import (
    CustomerActivateMembershipEmail,
)
from workshop_api.fitness.infrastructure.customer_entity import CustomerOrmModel
from workshop_api.fitness.infrastructure.customer_not_found_exception import (
    CustomerNotFoundException,
)
from workshop_api.fitness.infrastructure.customer_repository import CustomerRepository
from workshop_api.fitness.infrastructure.external_invoice_provider_client import (
    ExternalInvoiceProviderClient,
)
from workshop_api.fitness.infrastructure.in_memory_email_service import InMemoryEmailService
from workshop_api.fitness.infrastructure.membership_billing_reference_repository import (
    MembershipBillingReferenceRepository,
)
from workshop_api.fitness.infrastructure.membership_entity import (
    MembershipBillingReferenceOrmModel,
    MembershipOrmModel,
)
from workshop_api.fitness.infrastructure.membership_repository import MembershipRepository
from workshop_api.fitness.infrastructure.plan_not_found_exception import PlanNotFoundException
from workshop_api.fitness.infrastructure.plan_repository import PlanRepository


class MembershipService:
    def __init__(
        self,
        membership_repository: MembershipRepository,
        customer_repository: CustomerRepository,
        plan_repository: PlanRepository,
        billing_reference_repository: MembershipBillingReferenceRepository,
        email_service: InMemoryEmailService,
        billing_sender_email_address: str,
        external_invoice_provider_client: ExternalInvoiceProviderClient,
    ) -> None:
        self.membership_repository = membership_repository
        self.customer_repository = customer_repository
        self.plan_repository = plan_repository
        self.billing_reference_repository = billing_reference_repository
        self.email_service = email_service
        self.billing_sender_email_address = billing_sender_email_address
        self.external_invoice_provider_client = external_invoice_provider_client

    async def activate_membership(
        self,
        input_data: ActivateMembershipInput,
    ) -> ActivateMembershipResult:
        customer = self.customer_repository.find_by_id(input_data.customer_id)
        if customer is None:
            raise CustomerNotFoundException(
                f"Customer {input_data.customer_id} was not found"
            )

        plan = self.plan_repository.find_by_id(input_data.plan_id)
        if plan is None:
            raise PlanNotFoundException(f"Plan {input_data.plan_id} was not found")

        start_date = date.today()
        end_date = self._plus_months(start_date, plan.duration_in_months)
        customer_age = self._age_on(customer.date_of_birth, start_date)

        if customer_age < 18 and input_data.signed_by_custodian is not True:
            raise CustomerTooYoungException(
                "Customers younger than 18 require signedByCustodian=true"
            )

        membership = MembershipOrmModel(
            customer_id=str(input_data.customer_id),
            plan_id=str(input_data.plan_id),
            plan_price=int(Decimal(plan.price)),
            plan_duration=plan.duration_in_months,
            status="ACTIVE",
            start_date=start_date,
            end_date=end_date,
        )
        membership = self.membership_repository.save(membership)

        invoice_id = str(uuid.uuid4())
        invoice_due_date = date.today() + timedelta(days=30)
        external_invoice_id = await self.external_invoice_provider_client.create_membership_invoice(
            str(input_data.customer_id),
            membership,
            plan,
            invoice_due_date,
            invoice_id,
        )

        now = datetime.now(UTC)
        billing_reference = MembershipBillingReferenceOrmModel(
            membership_id=membership.id,
            external_invoice_id=external_invoice_id,
            external_invoice_reference=invoice_id,
            due_date=invoice_due_date,
            status="OPEN",
            created_at=now,
            updated_at=now,
        )
        billing_reference = self.billing_reference_repository.save(billing_reference)

        self._send_email(self._to_email(billing_reference, customer, membership))

        return ActivateMembershipResult(
            membership_id=uuid.UUID(str(billing_reference.membership_id)),
            customer_id=uuid.UUID(str(membership.customer_id)),
            plan_id=uuid.UUID(str(membership.plan_id)),
            plan_price=membership.plan_price,
            plan_duration=membership.plan_duration,
            status=membership.status,
            start_date=membership.start_date,
            end_date=membership.end_date,
            invoice_id=uuid.UUID(billing_reference.external_invoice_reference),
            external_invoice_id=billing_reference.external_invoice_id,
            invoice_due_date=billing_reference.due_date,
        )

    @staticmethod
    def _age_on(date_of_birth: date, checked_at: date) -> int:
        return checked_at.year - date_of_birth.year - (
            (checked_at.month, checked_at.day) < (date_of_birth.month, date_of_birth.day)
        )

    @staticmethod
    def _plus_months(start_date: date, months: int) -> date:
        total_month_index = start_date.month - 1 + months
        end_year = start_date.year + (total_month_index // 12)
        end_month = (total_month_index % 12) + 1
        end_day = min(start_date.day, calendar.monthrange(end_year, end_month)[1])
        return date(end_year, end_month, end_day)

    @staticmethod
    def _to_email(
        billing_reference: MembershipBillingReferenceOrmModel,
        customer: CustomerOrmModel,
        membership: MembershipOrmModel,
    ) -> CustomerActivateMembershipEmail:
        return CustomerActivateMembershipEmail(
            invoice_id=billing_reference.external_invoice_reference,
            invoice_due_date=billing_reference.due_date,
            email_address=customer.email_address,
            plan_price=membership.plan_price,
            email_template="""
To: {to}
From: {sender}
Subject: Your Membership Invoice {invoice_id}

Dear customer,

Thank you for your membership.

Please find your invoice details below:
Invoice ID: {invoice_id}
Amount Due: CHF {amount}
Due Date: {due_date}

Attachment: invoice-{invoice_id}.pdf

Kind regards,
Codeartify Billing
""",
        )

    def _send_email(self, email_details: CustomerActivateMembershipEmail) -> None:
        activation_email = email_details.email_template.strip().format(
            to=email_details.email_address,
            sender=self.billing_sender_email_address,
            invoice_id=email_details.invoice_id,
            amount=email_details.plan_price,
            due_date=email_details.invoice_due_date,
        )

        print(activation_email)
        self.email_service.send(activation_email)
